numba/targets/arrayobj.py

```diff
@@ -4857,21 +4859,189 @@ def np_sort_impl(a):
+        value = 3 -> (1, 0, :), value = 4 -> (1, 1, :), value = 5 -> (1, 2, :)
+
+    value should lie in range(np.prod(shape_tuple) // shape_tuple[axis]).
+    For this function to work, axis must be a literal integer.
+    """
+    if not isinstance(axis, types.IntegerLiteral):
+        raise RequireLiteralValue("axis argument must be a constant")
+
+    # Get the value of the axis constant.
+    axis_value = axis.literal_value
+    # The length of the indexing tuple to be output.
+    nd = len(shape_tuple)
+
+    # If the axis value is impossible for the given size array then
+    # just fake it like it was for axis 0.  This will stop compile errors
+    # when it looks like it could be called from array_sum_axis but really
+    # can't because that routine checks the axis mismatch and raise an
+    # exception.
+    if axis_value >= nd:
+        axis_value = 0
+
+    if axis_value < 0:
+        axis_value += nd
+
+    # Calculate the type of the indexing tuple.  All the non-axis
+    # dimensions have slice2 type and the axis dimension has int type.
+    before = axis_value
+    after = nd - before - 1
+
+    types_list = []
+    types_list += [types.intp] * before
+    types_list += [types.slice2_type]
+    types_list += [types.intp] * after
+
+    # Creates the output type of the function.
+    tupty = types.Tuple(types_list)
+    # Defines the signature of the intrinsic.
+    function_sig = tupty(shape_tuple, value, axis)
+
+    def codegen(cgctx, builder, signature, args):
+        lltupty = cgctx.get_value_type(tupty)
+        # Create an empty indexing tuple.
+        tup = cgutils.get_null_value(lltupty)
+
+        # We only need value of the axis dimension here.
+        # The rest are constants defined above.
+        [shape_arg, value_arg, _] = args
+
+        def create_full_slice():
+            return slice(None, None)
+
+        for i in range(nd - 1, axis_value, -1):
+            dim = builder.extract_value(shape_arg, i)
+            modulo = builder.srem(value_arg, dim)
+            tup = builder.insert_value(tup, modulo, i)
+            value_arg = builder.sdiv(value_arg, dim)
+
+        # compile and call create_full_slice
+        slice_data = cgctx.compile_internal(builder, create_full_slice,
+                                            types.slice2_type(),
+                                            [])
+        tup = builder.insert_value(tup, slice_data, axis_value)
+
+        for i in range(axis_value - 1, -1, -1):
+            dim = builder.extract_value(shape_arg, i)
+            modulo = builder.srem(value_arg, dim)
+            tup = builder.insert_value(tup, modulo, i)
+            value_arg = builder.sdiv(value_arg, dim)
+
+        return tup
+
+    return function_sig, codegen
+
+
+@overload_method(types.Array, "argsort")
+@overload(np.argsort)
+def np_argsort(a, axis=-1, kind="quicksort"):
+    # Unpack TypeRef
+    axis = getattr(axis, "instance_type", axis)
+    kind = getattr(kind, "instance_type", kind)
```

Comment:
You don't need to unpack `TypeRef`, which only occurs if there is code that passes in a typeobj into the args; like `np.argsort(a, axis=numba.types.Integer)`.